### PR TITLE
Override tokenize.ensure-deterministic config flag

### PR DESCRIPTION
**Pull request by crusaderky**
- Add optional parameter to `tokenize(ensure_deterministic=True)` to force raising exception
- Can now run `test_tokenize.py` with either value for the config variable
- Subclass RuntimeError to restrict exception catching (we're calling pickle internally, which is very complex)
- Better tests for the `_seen` variable

### CODE REVIEW

File: dask/base.py
Comment (fjetter) [inline dask/base.py:1055]: This seems overly complex. Why isn't a single `ContextVar` sufficient?

Comment (fjetter) [reply dask/base.py:1055]: I only noticed now that this was already introduced in https://github.com/dask/dask/pull/10876

File: dask/base.py
Comment (fjetter) [inline dask/base.py:1373]: This feels like an unrelated change. While the functionality is similar, the raised exception is a little different

File: dask/tests/test_tokenize.py
Comment (fjetter) [inline dask/tests/test_tokenize.py:62]: I'm not sure if this is what you are referring to in the comment above but don't we want this to be equal as before?
```suggestion
        assert tok2 == tok3 == before
```

File: dask/tests/test_tokenize.py
Comment (fjetter) [inline dask/tests/test_tokenize.py:167]: Probably doesn't matter much but it's more likely to be an instance
```suggestion
    x = np.array(["a", None, NeedsCloudPickle()], dtype=object)
```

File: dask/tests/test_tokenize.py
Comment (fjetter) [inline dask/tests/test_tokenize.py:293]: This test is about ufuncs. Asserting on this ctxvar feels odd.

File: dask/tests/test_tokenize.py
Comment (fjetter) [inline dask/tests/test_tokenize.py:324]: why remove that?

File: dask/base.py
Comment (crusaderky) [reply dask/base.py:1373]: Blocked by and incorporates https://github.com/dask/dask/pull/10883.
To review this PR before the above is merged, just look at the last commit.

File: dask/tests/test_tokenize.py
Comment (crusaderky) [reply dask/tests/test_tokenize.py:324]: This PR incorporates https://github.com/dask/dask/pull/10883.
To review this PR before the above is merged, just look at the last commit.
This comment is about a change in #10883; please keep the critique there.

File: dask/tests/test_tokenize.py
Comment (crusaderky) [reply dask/tests/test_tokenize.py:62]: This PR incorporates https://github.com/dask/dask/pull/10883.
To review this PR before the above is merged, just look at the last commit.
This comment is about a change in #10883; please keep the critique there.

File: dask/base.py
Comment (crusaderky) [reply dask/base.py:1373]: This PR incorporates https://github.com/dask/dask/pull/10883.
To review this PR before the above is merged, just look at the last commit.
This comment is about a change in #10883; please keep the critique there.

File: dask/tests/test_tokenize.py
Comment (crusaderky) [reply dask/tests/test_tokenize.py:293]: The test may fail if I were to tamper with the ctxvar without using the helper function `_maybe_raise_nondeterministic`

Comment (hendrikmakait) [reply dask/tests/test_tokenize.py:293]: This feels like it would be useful as a comment.

File: dask/tests/test_tokenize.py
Comment (hendrikmakait) [inline dask/tests/test_tokenize.py:318]: Should this be nested within the block above?

File: dask/tests/test_tokenize.py
Comment (hendrikmakait) [inline dask/tests/test_tokenize.py:287]: Should this be nested within the block above?

File: dask/tests/test_tokenize.py
Comment (hendrikmakait) [inline dask/tests/test_tokenize.py:374]: Should this be nested within the block above?

File: dask/tests/test_tokenize.py
Comment (crusaderky) [reply dask/tests/test_tokenize.py]: I reworked it.

File: dask/tests/test_tokenize.py
Comment (crusaderky) [reply dask/tests/test_tokenize.py:299]: No. These 3 tests use the same pattern:

```python
with dask.config.set({"tokenize.ensure-deterministic": False}):
    tokenize(...)
with dask.config.set({"tokenize.ensure-deterministic": True}):
    tokenize(...)
tokenize(..., ensure_determinsitic=False)
tokenize(..., ensure_deterministic=True)
```
This makes them impervious to whatever the config is set to globally.

File: dask/tests/test_tokenize.py
Comment (crusaderky) [reply dask/tests/test_tokenize.py:326]: same answer as above

File: dask/tests/test_tokenize.py
Comment (crusaderky) [reply dask/tests/test_tokenize.py:378]: same answer as above

**Comment by crusaderky** [comment]
@hendrikmakait @fjetter all comments have been addressed

File: dask/tests/test_tokenize.py
Comment (hendrikmakait) [reply dask/tests/test_tokenize.py:299]: My bad, I attributed the inline comment only to this test case, not the one below.

Comment (hendrikmakait) [reply dask/tests/test_tokenize.py:299]: I also wasn't sure if this should test overriding "whatever" the config is set to globally or overriding _any possible_ config value.

**Review by hendrikmakait — APPROVED** [review]
Thanks, @crusaderky! LGTM assuming CI is green

**Pull request merged**
